claat/serve.go

```diff
@@ -0,0 +1,221 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"archive/zip"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+)
+
+var deps = map[string]string{
```

Comment:
This should probably be loader from an external file, ideally which is auto-generated by make.

Reply:
I would just grab and parse bower.json file from codelab-components:
https://github.com/googlecodelabs/codelab-components/blob/master/bower.json
It already contains all the deps we need and has nearly identical structure to what you already have here. Just ignore the rest of the stuff you'll find in bower.json.

Reply:
That's where I started but the problem with that is I don't pick up the dependency tree.

Reply:
Right, so instead of calling `downloadFile(...)` you'll have `fetchRepo(destDir, "githuborg/repo")` which checks whether this package already exists on dist and, if not, again looks up `bower.json` in that repo and calls `fetchRepo` recursively. So, very simple, just one more function which ignores the dependency tree resolution and picks up whichever first version was downloaded.

Reply:
Here's what I'm seeing (in pseudocode):
fetchRepo(destDir, url)
- parse url into repoName and version
- get github.com/repoName/archive/version/master.zip
- unzip downloaded file into destDir
- if unzipped archive contains a bower.json, parse it, and for each dependency depStr in that json doc:
    - recursively call fetchRepo(destDir, depStr)

With that function, I can snarf everything I need by calling fetchRepo(destDir, "googelcodelabs/codelab-components#1.0.0").

wdyt?

Reply:
Exactly what I had in mind, too.
- the `fetchRepo` should return immediately if the directory already exists
- the URL is actually of the form `https://github.com/googlecodelabs/codelab-components/archive/<version>.zip`. Note there's no `master`. I gave you incorrect one before